RBD fscrypt support

irq0 opened the pull request:
Add fscrypt integration with the Ceph CSI KMS. Supports ext4 on RBD. Snapshots are supported as well. 
Implementation is based on the design doc https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/cephfs-fscrypt.md.

Testing requires kernel support (see https://github.com/google/fscrypt#runtime-dependencies). Minikube needs https://github.com/kubernetes/minikube/pull/14783 to run the e2e test suite.


Commits are ordered as follows : 
1. utils and refactoring
2. general fscrypt integration
3. fscrypt into RBD integration
4. e2e testing

irq0 committed:
util: Make encryption passphrase size a parameter

fscrypt support requires keys longer than 20 bytes. As a preparation,
make the new passphrase length configurable, but default to 20 bytes.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
kms: Add GetSecret() to metadata KMS

Add GetSecret() to allow direct access to passphrases without KDF and
wrapping by a DEKStore.

This will be used by fscrypt, which has its own KDF and wrapping. It
will allow users to take a k8s secret, for example, and use that
directly as a password in fscrypt.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
kms: Add basic GetSecret() test

Add rudimentary test to ensure that we can get a valid passphrase from
the GetSecret() feature

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rbd: Rename encryption to blockEncryption prep for fscrypt

In preparation of fscrypt support for RBD filesystems, rename block
encryption related function to include the word 'block'. Add struct
fields and IsFileEncrypted.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rbd: fscrypt file encryption support

Integrate basic fscrypt functionality into RBD initialization. To
activate file encryption instead of block introduce the new
'encryptionType' storage class key.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
vendor: vendor fscrypt integration dependencies

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rbd: Document new encryptionType storage class example

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Add fscrypt on rbd helper

Add validation functions for fscrypt on RBD volumes

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
util: Add util to fetch encryption type from vol options

Fetch encryption type from vol options. Make fallback type
configurable to support RBD (default block) and Ceph FS (default file)

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Add helper to run encryption tests on block and file

Add a `By` wrapper to parameterize encryption related test functions
and run them on both block and file encryption

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Run encryption related tests on file and block type

Replace `By` with `ByFileAndBlockEncryption` in all encryption related
tests to parameterize them to file and block encryption.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: Unlock: Fetch keys early

Fetch keys from KMS before doing anything else. This will catch KMS
errors before setting up any fscrypt metadata.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: Fetch passphrase when keyFn is invoked not created

Fetch password when keyFn is invoked, not when it is created. This
allows creation of the keyFn before actually creating the passphrase.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
journal: Store encryptionType in Config struct

Add encryptionType next to kmsID to support both block and file
encryption.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
cephfs: Add placeholder journal fscrypt support

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: Update mount info before create context

NewContextFrom{Mountpoint,Path} functions use cached
`/proc/self/mountinfo` to find mounted file systems by device ID.
Since we run fscrypt as a library in a long-lived process the cached
information is likely to be stale. Stale entries may map device IDs to
mount points of already destroyed RBDs and fail context creation.
Updating the cache beforehand prevents this.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rbd: Handle encryption type default at a more meaningful place

Different places have different meaningful fallback. When parsing
from user we should default to block, when parsing stored config we
should default to invalid and handle that as an error.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rbd: support file encrypted snapshots

Support fscrypt on RBD snapshots

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Add PVC validator to ByFileAndBlockEncryption

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: Use constant protector name

Use constant protector name 'ceph-csi' instead of constant prefix
concatenated with the volume ID. When cloning volumes the ID changes
and fscrypt protected directories become inunlockable due to the
protector name change

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: fsync encrypted dir after setting policy [workaround]

Revert once our google/fscrypt dependency is upgraded to a version
that includes https://github.com/google/fscrypt/pull/359 gets accepted

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: fscrypt integration

Integrate google/fscrypt into Ceph CSI KMS and encryption setup. Adds
dependencies to google/fscrypt and pkg/xattr. Be as generic as
possible to support integration with both RBD and Ceph FS.

Add the following public functions:

InitializeNode: per-node initialization steps. Must be called
before Unlock at least once.

Unlock: All steps necessary to unlock an encrypted directory including
setting it up initially.

IsDirectoryUnlocked: Test if directory is really encrypted

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: Determine best supported fscrypt policy on node init

Currently fscrypt supports policies version 1 and 2. 2 is the best
choice and was the only choice prior to this commit. This adds support
for kernels < 5.4, by selecting policy version 1 there.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rbd: Add volume journal encryption support

Add fscrypt support to the journal to support operations like
snapshotting.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
kms: testing: add KMS test dummy registry

Add registry similar to the providers one. This allows testers to
add and use GetKMSTestDummy() to create stripped down provider
instances suitable for use in unit tests.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Add encrypted PVC with default settings test

Add test that enables encryption with default type. Check that we set
up block encryption.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
Was https://github.com/ceph/ceph-csi/pull/3158 

As this is already pretty large, I left out the Ceph FS bits for now. (They are on https://github.com/irq0/ceph-csi/tree/wip/fscrypt-go-rbd-cephfs)

nixpanic commented on internal/util/crypto.go:
```diff
@@ -80,6 +80,50 @@ func FetchEncryptionKMSID(encrypted, kmsID string) (string, error) {
+type EncryptionType int
+
+const (
+	EncryptionTypeInvalid EncryptionType = iota
```
I guess that documenting the exported constants is required, CI might fail because of it.

irq0 replied:
:+1:

nixpanic commented on internal/util/crypto.go:
```diff
@@ -80,6 +80,50 @@ func FetchEncryptionKMSID(encrypted, kmsID string) (string, error) {
+	EncryptionTypeFile                   = iota
+)
+
+func ParseEncryptionType(typeStr string) EncryptionType {
```
unit tests for these functions would be good

irq0 replied:
:+1:

Commit:
kms: Add GetSecret() to KMIP KMS

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Apply formatting to rbd suite and helper

Apply formatting for previous changes separately to make the commit
diffs easier to read.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
The last push rebased on current develop branch and addressed the linter errors. The minikube tests on CI will again fail, as minikube doesn't have ext4 fscrypt support.

The tests failing with timeouts raise a question though: Maybe not having kernel support should return a not retryable error instead of an internal error.
Maybe codes.Unimplemented (spec: https://github.com/container-storage-interface/spec/blob/master/spec.md#error-scheme)? Thaughts?

Commit:
scripts: Add env to set minikube iso url

Make iso url configurable to use pre-release minikube images or
local-built (file://)

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Use utilEncryptionType instead of string in rbd suite

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
:partying_face: the e2e suites passed. The linter finds a line that is a bit too long. I already fixed it in my local branch. I suggest I push the fix with the first round of review changes. 

From my side this is ready to review.

nixpanic commented on internal/kms/dummy.go:
```diff
@@ -0,0 +1,71 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
```
as these functions are test-only, can they be placed in `dummy_test.go` or `utils_test.go` so that they only get compiled/linked while running the tests? Or maybe there is an automatic BuildConstraint while running `go test`? https://pkg.go.dev/cmd/go#hdr-Build_constraints might have some details (but I did not find it that quickly).

irq0 replied:
Did not work, because getsecret_test.go needs those functions. More in next comment

nixpanic commented on internal/util/getsecret_test.go:
```diff
@@ -0,0 +1,52 @@
+		if provider.CreateTestDummy == nil {
+			continue
+		}
+		kms := kmsapi.GetKMSTestDummy(provider.UniqueID)
```
ah, right, this is a cross-package testing function. Placing it in a test-only compile-time thing may not work (to come back on my previous comment).

irq0 replied:
I initially tried to make dummy.go test only, but it didn't work. I suspect that it was due to the package split.

dummy.go and getsecret_test.go are a bit suboptimal split, but I ran into cyclic include problems with the test in the kms package.

nixpanic commented on internal/util/fscrypt/fscrypt.go:
```diff
@@ -13,6 +13,11 @@ limitations under the License.
+#include <linux/fs.h>
+*/
+import "C"
+
```
this commit added the `#include <linux/fs.h>`, but I do not see it used anywhere (within this commit).

irq0 replied:
good catch! moved to the fscrypt integration commit, where it is required for `unix.Syscall(unix.SYS_IOCTL, file.Fd(), unix.FS_IOC_GETFLAGS, ...)`

nixpanic commented on internal/util/fscrypt/fscrypt.go:
```diff
@@ -111,6 +111,21 @@ func createKeyFuncFromVolumeEncryption(
 	return keyFunc, nil
 }
 
+func fsyncEncryptedDirectory(dirPath string) error {
```
can you leave a `TODO` comment for this function, pointing to google/fscrypt#359 ?

irq0 replied:
:+1:

nixpanic commented on internal/rbd/encryption.go:
```diff
@@ -278,25 +330,27 @@ func (ri *rbdImage) initKMS(ctx context.Context, volOptions, credentials map[str
 	encrypted, ok = volOptions["encrypted"]
 	if !ok {
-		return "", nil
+		return "", util.EncryptionTypeInvalid, err
```
how about returning `"", util.EncryptionTypeNone, nil` as this is a success case? Encryption is not enabled for this volume, which is valid.

Also, returning `err` (which is `nil` at the moment) is dangerous, it can be that something else is added to this function later on, and `err` may not stay `nil` in that case. To improve maintainability, explicit `nil` returns on success have my preference.

irq0 replied:
:+1:

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -433,6 +434,12 @@ func (ns *NodeServer) stageTransaction(
 
+	if volOptions.isFileEncrypted() {
+		if err = fscrypt.InitializeNode(ctx); err != nil {
+			return transaction, err
```
Does `InitializeNode()` return a clear error? If not, add some more context with `fmt.Errorf(...)` please.

irq0 replied:
:+1:

nixpanic commented on internal/rbd/encryption.go:
```diff
@@ -262,14 +306,22 @@ func (rv *rbdVolume) openEncryptedDevice(ctx context.Context, devicePath string)
 	}
 
-	err = ri.configureBlockDeviceEncryption(kmsID, credentials)
+	switch encType {
```
I never liked checking `kmsID == ""` above (or anywhere else). If there is a `util.EncryptionTypeNone` value, it can be included in the switch/case. That would be a nice cleanup!

irq0 replied:
:+1:

nixpanic commented on examples/rbd/storageclass.yaml:
```diff
@@ -108,6 +108,12 @@ parameters:
+   # above. Defaults to "block". Valid values are: "file": Enable file
+   # encryption on the mounted filesystem; "block": Encrypt RBD block
+   # device
+   # encryptionType: "block"
```
On RBD, both could be enabled. I think that is (mostly) possible with the implementation as well? Should configuring that be prevented?

irq0 replied:
In theory file and block encryption are orthogonal. At one point I considered a "both" option, but ultimately rejected it. The complexity is IMHO not worth it. Most of the complexity would be in configuration and KMS init.

With two encryption providers: Use a single or individual KMS? Share secrets? Namespace secrets? How to configure multiple KMS properly?

Considering that Ceph FS will only ever have file encryption, there being little benefit in stacking filesystem and block crypto and the complexity in KMS initialization and configuration I think either or is the way to go.

nixpanic replied:
I do not think there is a need to support both on the same volume. Specially if it requires more work! Just note somewhere that the choice is _none_, _block_ or _file_.

irq0 replied:
I'll try to make the comment more clear. The choice is actually "block", "file" and not setting the key. If unset it defaults to block internally. This ensures pre file encryption configs without encryptionType continue to work (there is an e2e test that covers this).

nixpanic commented on e2e/rbd.go:
```diff
@@ -232,6 +232,15 @@ func checkClusternameInMetadata(f *framework.Framework, ns, pool, image string)
 }
 
+// ByFileAndBlockEncryption wraps ginkgo's By to run the test body using file and block encryption specific validators.
+func ByFileAndBlockEncryption(
```
this is much appreciated, less copy/pasting all around! We should do something like this more often to cleanup the e2e code.

nixpanic commented on build.env:
```diff
@@ -39,6 +39,7 @@ HELM_VERSION=v3.9.2
 
 # minikube settings
 MINIKUBE_VERSION=v1.26.1
+MINIKUBE_ISO_URL=https://storage.googleapis.com/minikube-builds/iso/14783/minikube-v1.26.1-1661377864-14783-amd64.iso
```
A `FIXME: ` comment above this URL would be good, refer to the PR that contains the fscrypt changes.

irq0 replied:
:+1:

nixpanic reviewed (changes requested):
Looks really good! Sorry for needing a long time to review. I have come across only a few minor things. Thanks for continuing with this huge contribution!

Commit:
rbd: Add context to fscrypt errors

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
util: Add EncryptionTypeNone and unit tests

Add type none to distinguish disabled encryption (positive result)
from invalid configuration (negative result).

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rbd: Use EncryptionTypeNone

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
fscrypt: fix metadata directory permissions

Call Mount.Setup with SingleUserWritable constant instead of 0o755,
which is silently ignored and causes the /.fscrypt/{policy,protector}/
directories to have mode 000.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
The last push addresses the review comments and adds a small fix to the fscrypt integration code (fscrypt: fix metadata directory permissions). It didn't cause problems on rbd+ext4, but on cephfs.


ci/centos/k8s-e2e-external-storage/1.23, ci/centos/mini-e2e-helm/k8s-1.22 failed in the 'reserve bare-metal machine' stage. Is there a way to trigger them individually?

irq0 commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

irq0 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

irq0 commented:
> ci/centos/k8s-e2e-external-storage/1.23, ci/centos/mini-e2e-helm/k8s-1.22 failed in the 'reserve bare-metal machine' stage. Is there a way to trigger them individually?

Found it :)

nixpanic commented:
> > ci/centos/k8s-e2e-external-storage/1.23, ci/centos/mini-e2e-helm/k8s-1.22 failed in the 'reserve bare-metal machine' stage. Is there a way to trigger them individually?
> 
> Found it :)

Not all CI jobs need to pass while this is getting reviewed. Once the PR is approved, we'll ask Mergify to rebase it and all CI jobs need to run+pass at that time, just before it gets merged.

nixpanic reviewed (dismissed):
Thanks! Looks pretty good to me, lets get this in soon and get it tested :+1:

irq0 commented:
> Thanks! Looks pretty good to me, lets get this in soon and get it tested +1

That would be awesome. Anything I can help with to get this in soon? 
I have also a follow up with Ceph FS fscrypt support in the pipeline. That one won't be as big :) 

The last push fixed the go.sum mergify complained about.

irq0 commented:
/retest ci/centos/mini-e2e/k8s-1.24

Madhu-1 commented on internal/kms/dummy.go:
```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2022 SUSE LLC.
```
why SUSE LLC?

irq0 replied:
I work for SUSE

Madhu-1 replied:
Agree on above thoughts :+1: is it something you want to consider, we have someone contribution in cephcsi repo everyone uses ceph-csi authors as its a community project we dont have individual copywrite per company/contributors.

irq0 replied:
OK. I'll update the files

Madhu-1 commented on internal/util/crypto.go:
```diff
@@ -80,6 +80,68 @@ func FetchEncryptionKMSID(encrypted, kmsID string) (string, error) {
 
+type EncryptionType int
+
+const (
+	// EncryptionTypeInvalid signals invalid or unsupported configuration.
+	EncryptionTypeInvalid EncryptionType = iota
+	// EncryptionTypeNone disables encryption.
+	EncryptionTypeNone = iota
+	// EncryptionTypeBlock enables block encryption.
+	EncryptionTypeBlock = iota
+	// EncryptionTypeBlock enables file encryption (fscrypt).
+	EncryptionTypeFile = iota
+)
```
```suggestion
const (
	// EncryptionTypeInvalid signals invalid or unsupported configuration.
	EncryptionTypeInvalid EncryptionType = iota
	// EncryptionTypeNone disables encryption.
	EncryptionTypeNone
	// EncryptionTypeBlock enables block encryption.
	EncryptionTypeBlock
	// EncryptionTypeBlock enables file encryption (fscrypt).
	EncryptionTypeFile
)
```

irq0 replied:
Done

Madhu-1 commented on internal/util/crypto.go:
```diff
@@ -80,6 +80,68 @@ func FetchEncryptionKMSID(encrypted, kmsID string) (string, error) {
+	EncryptionTypeFile = iota
+)
+
+const (
```
define type and use it as type const?

irq0 replied:
See below

Madhu-1 commented on internal/util/crypto.go:
```diff
@@ -80,6 +80,68 @@ func FetchEncryptionKMSID(encrypted, kmsID string) (string, error) {
+	case EncryptionTypeNone:
+		return ""
+	case EncryptionTypeInvalid:
+		return "INVALID"
+	default:
+		return "UNKNOWN"
```
define this as typed const and return it?

irq0 replied:
The string representation is only used when printing the EncryptionType type. Using a typed const would mean that we need a lot of conversions back to string across the codebase. All checks are done on the EncryptionType. The reason EncryptionTypeBlockString and EncryptionTypeFileString exist is a linter error about the repeated use of what appears to be string constants. Looking at them they are better not exported.

Madhu-1 commented on internal/util/fscrypt/fscrypt.go:
```diff
@@ -0,0 +1,444 @@
+/*
+Copyright 2022 SUSE LLC.
```
why SUSE LLC?

Madhu-1 commented on internal/util/fscrypt/fscrypt.go:
```diff
@@ -0,0 +1,444 @@
+	defer dir.Close()
+
+	err = dir.Sync()
+	if err != nil {
+		return err
+	}
+
+	return nil
```
just return err?

irq0 replied:
Good catch!

Madhu-1 commented on internal/util/fscrypt/fscrypt.go:
```diff
@@ -0,0 +1,444 @@
+	defer func() {
+		err = policy.Lock()
+		if err != nil {
+			log.ErrorLog(ctx, "fscrypt: failed to lock policy after use: %w", err)
```
log with '%v` instead of `%w`

irq0 replied:
Done

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -176,6 +177,10 @@ else
     DISK_CONFIG=""
 fi
 
+if [[ -n "${MINIKUBE_ISO_URL}" ]]; then
+    EXTRA_CONFIG="${EXTRA_CONFIG} --iso-url ${MINIKUBE_ISO_URL}"
+fi
```
Need to check this works with podman driver as we are moving to podman driver to create kubernetes cluster

nixpanic replied:
This won't be used when minikube uses podman. The VM where the job is running will need to have a kernel with the right options now :cry: CentOS Stream 8 is used, we might want to move the CI to CentOS Stream 9 (we're running in the CentOS CI).

nixpanic replied:
we can consider running a daily job that runs on a Amazon bare-metal system and deploy minikube with its VM there. But for regular CI jobs, we'll rely on Amazon VMs for now.

irq0 replied:
I just started a CentOS 8 Stream VM, it has kernel 4.18. That is a bit too old to properly run fscrypt. Ideally this would be >= 5.4 [0].

Something bare metal would be awesome! The Ceph FS fscrypt support would  require something even newer..

[0] https://github.com/google/fscrypt#runtime-dependencies

irq0 replied:
How to move forward? Feature flag the file encryption tests?

Madhu-1 reviewed (commented):
some nits, changes LGTM

irq0 commented:
The latest push addresses the comments from @Madhu-1 and is rebased to current develop.

nixpanic commented:
I'm a little sad about it, but with the move to AWS EC2 VMs, the change to the minikube kernel is now unused. Is it possible to add a `testFSCrypt` option in `e2e/e2e_test.go` and have it disabled by default? In the future, we should have a regular run on bare-metal systems again, and we can enable it for those jobs. It would be similar to the `testNBD` option I added earlier this week.

Commit:
e2e: Feature flag RBD fscrypt tests (default disabled)

Add test-rbd-fscrypt feature flag to e2e suite. Default disabled as
the current CI system's kernel doesn't have the required features
enabled.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
> I'm a little sad about it, but with the move to AWS EC2 VMs, the change to the minikube kernel is now unused. Is it possible to add a `testFSCrypt` option in `e2e/e2e_test.go` and have it disabled by default? In the future, we should have a regular run on bare-metal systems again, and we can enable it for those jobs. It would be similar to the `testNBD` option I added earlier this week.

A bit unfortunate to not have it in the standard CI test set, yes :(

Having fscrypt in minikube is still a very good thing. Without it testing would require a custom minikube build. Looking further the way fscrypt is enabled in the kernel also means that CephFS fscrypt support will come automatically eventually to minikube. 


The last push adds a commit with the feature flag. I called it testRBDFscrypt / test-rbd-fscrypt, because the Ceph FS integration already has a testCephFSFscrypt / test-cephfs-fscrypt flag.

Rakshith-R commented on internal/util/fscrypt/fscrypt.go:
```diff
@@ -0,0 +1,439 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
```
```suggestion
/*
Copyright 2022 The Ceph-CSI Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
```

Can you please add the newlines too similar to  other existing files.

Rakshith-R commented on internal/util/fscrypt/fscrypt.go:
```diff
@@ -0,0 +1,439 @@
+	"time"
+	"unsafe"
+
+	fscryptactions "github.com/google/fscrypt/actions"
+	fscryptcrypto "github.com/google/fscrypt/crypto"
+	fscryptfilesystem "github.com/google/fscrypt/filesystem"
+	fscryptmetadata "github.com/google/fscrypt/metadata"
+	"github.com/pkg/xattr"
+	"golang.org/x/sys/unix"
+
+	"github.com/ceph/ceph-csi/internal/kms"
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
```
```suggestion
	"github.com/ceph/ceph-csi/internal/kms"
	"github.com/ceph/ceph-csi/internal/util"
	"github.com/ceph/ceph-csi/internal/util/log"

	fscryptactions "github.com/google/fscrypt/actions"
	fscryptcrypto "github.com/google/fscrypt/crypto"
	fscryptfilesystem "github.com/google/fscrypt/filesystem"
	fscryptmetadata "github.com/google/fscrypt/metadata"
	"github.com/pkg/xattr"
	"golang.org/x/sys/unix"
```

`golang lib - internal - third party `
